New Checkout Drop-in v2.0.0-beta1, beta2 y beta3 (#289)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/checkout/error-handling.mdx

---
title: Error Handling

> description: Learn about on how Checkout drop-in handles errors.

Reviewer: ```suggestion
description: Learn about how the Checkout drop-in handles errors that occur during the checkout process.
```

---

Errors ocurred during the checkout process should be caught and logged providing clear context for quick resolution to prevent unnecessary propagation and have a better user experience and debugging. Checkout drop-in must implement an error handling mechanism to improve observability and debugging capabilities.

It's critical to resolve the errors to avoid incongruent situations and clearly inform the user of what happened to avoid data inconsistencies between the local application and the backend, which may in turn resulted in an incorrect order.

## Generic strategy

In general, most issues will be due to errors caused by API calls. Therefore, it is needed to focus on how these errors propagate upward from dispatching the API calls to the user interface and how they are presented to the user in a friendly manner in different scenarios. To do this, each container needs a centralized error handling system, capturing the errors as they occur so we can have control on them, and decide what are the next actions to perform.

## "Optimistic" UI updates with rollback pattern

> The system should work using "Optimistic" UI updates with rollback concept. It’s a technique used to improve the user experience by making the application feel like its more responsive.

Reviewer: ```suggestion
The system should work using "Optimistic" UI updates with rollback concept. It is a technique used to improve the user experience by making the application feel like it is more responsive.
```

In an optimistic update the UI behaves as though a change was successfully completed before receiving confirmation from the backend that it actually was - it is being optimistic that it will eventually get the confirmation rather than an error. This allows for a more responsive user experience.

> In this scenario, when the user performs an action that changes the state, that action sends information to the backend and the UI is optimistically updated to reflect the change in state. We say “optimistically” because we are creating this UI change with hope and optimism that the change in state will be accepted by the backend (if we wait for the backend to tell us that the user action was successful before conducting a UI change, it would take an unacceptably long time and the user experience would not be great).

Reviewer: ```suggestion
In this scenario, when the user performs an action that changes the state, that action sends information to the backend, and the user interface (UI) is optimistically updated to reflect the change. We refer to this process as "optimistic" because we are making the UI change with the belief that the backend will accept the state change. If we waited for confirmation from the backend that the user action was successful before updating the UI, it would take too long, negatively impacting the user experience.
```
